### PR TITLE
Adding Tests and isolating JWT

### PR DESCRIPTION
**Pull request by lgaches**
- Isolate JWT on its own module
- Add some tests
- Alert and APS payload complete
- Run test on CircleCI

### CODE REVIEW

File: Sources/NIOAPNS/APNSRequestEncoder.swift [hunk 1]
```diff
@@ -22,17 +23,19 @@ internal final class APNSRequestEncoder<Notification>: ChannelOutboundHandler
     let configuration: APNSConfiguration
     let deviceToken: String
     
+    let encoder = JSONEncoder()
+
     init(deviceToken: String, configuration: APNSConfiguration) {
         self.configuration = configuration
-        self.deviceToken = deviceToken
+        self.deviceToken = deviceToken
```
Comment (naknut) [inline Sources/NIOAPNS/APNSRequestEncoder.swift]: I think you accidentally added a white space at the end here.

File: Sources/NIOAPNSJWT/DataSigner.swift [hunk 2]
```diff
@@ -10,7 +10,7 @@ import CAPNSOpenSSL
 public class DataSigner: APNSSigner {
     private let opaqueKey: OpaquePointer
 
-    public init(data: Data) throws {
+    public init(data: Data) throws {
```
Comment (naknut) [inline Sources/NIOAPNSJWT/DataSigner.swift]: An other accidental whitespace.

File: Sources/NIOAPNS/APNSRequestEncoder.swift [hunk 3]
```diff
@@ -22,17 +23,19 @@ internal final class APNSRequestEncoder<Notification>: ChannelOutboundHandler
...
+        self.deviceToken = deviceToken        
     }
     
     /// See `ChannelOutboundHandler.write(context:data:promise:)`.
     func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
         let req: Notification = unwrapOutboundIn(data)
         let data: Data
         do {
-            data = try JSONEncoder().encode(req)
+            data = try encoder.encode(req)
```
Comment (naknut) [inline Sources/NIOAPNS/APNSRequestEncoder.swift]: Is there a reason this is moved to a global property?

**Comment by kylebrowning** [comment]
Thanks for all this hard work. This is nice!

File: Sources/NIOAPNSJWT/DataSigner.swift [hunk 4]
```diff
@@ -40,29 +40,17 @@ public class DataSigner: APNSSigner {
...
             throw APNSSignatureError.invalidASN1
         }
-        
-        // Force unwrap because guard protects us.
-        return Data(bytesNoCopy: derEncodedSignature!,
-                    count: Int(derLength),
-                    deallocator: .custom({ pointer, length in CRYPTO_free(pointer) }))
-    }
-
-    public func verify(digest: Data, signature: Data) -> Bool {
```
Comment (kylebrowning) [inline Sources/NIOAPNSJWT/DataSigner.swift]: Why do we remove the verify when it could have been used in tests?

File: Sources/CAPNSOpenSSL/shim.h [hunk 5]
```diff
@@ -8,13 +8,6 @@
 #include <openssl/opensslv.h>
 #include <openssl/crypto.h>
 
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || !defined(LIBRESSL_VERSION_NUMBER)
```
Comment (kylebrowning) [inline Sources/CAPNSOpenSSL/shim.h]: This is needed on some linux systems with openssl below 1.1.

Why did you remove it? Because you couldn't compile on a Mac I presume? 

Checkout this thread, and this comment, https://github.com/kylebrowning/swift-nio-http2-apns/issues/9#issuecomment-479084231

File: Tests/NIOAPNSJWTTests/JWTTests.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        let teamID = "8RX5AF8F6Z"
+        let keyID = "9N8238KQ6Z"
+        let date = Date()
+
+        guard let pk = appleECP8PrivateKey.data(using: .utf8) else { XCTFail("Can't convert private key string to Data"); return}
+        let signer = try SigningMode.data(data: pk)
+        let jwt = JWT(keyID: keyID, teamID: teamID, issueDate: date, expireDuration: 10.0)
+        let token = try jwt.sign(with: signer)
+
+        let part = token.split(separator: ".")
```
Comment (kylebrowning) [inline Tests/NIOAPNSJWTTests/JWTTests.swift]: Surely we could use the `signer.verify` here?

File: Sources/NIOAPNSJWT/DataSigner.swift (hunk 4 above)
Comment (lgaches) [reply Sources/NIOAPNSJWT/DataSigner.swift]: Because it's useless. And I had compile error on Linux

File: Sources/CAPNSOpenSSL/shim.h (hunk 5 above)
Comment (lgaches) [reply Sources/CAPNSOpenSSL/shim.h]: Yes I had a hard time with it on Mac

File: Tests/NIOAPNSJWTTests/JWTTests.swift (hunk 6 above)
Comment (lgaches) [reply Tests/NIOAPNSJWTTests/JWTTests.swift]: Yes was thinking the same

File: Sources/NIOAPNS/APNSRequestEncoder.swift (hunk 3 above)
Comment (lgaches) [reply Sources/NIOAPNS/APNSRequestEncoder.swift]: nope I will revert that

File: Sources/NIOAPNSJWT/DataSigner.swift (hunk 4 above)
Comment (kylebrowning) [reply Sources/NIOAPNSJWT/DataSigner.swift]: I suppose we could remove it since it is not used anywhere. If anyone needs it they can always used the custom version. 

Out of curiosity, if you find time, let me know what the version of OpenSSL was.

File: Sources/CAPNSOpenSSL/shim.h (hunk 5 above)
Comment (lgaches) [reply Sources/CAPNSOpenSSL/shim.h]: I reverted my change and added a comment about it on the README

Comment (kylebrowning) [reply Sources/CAPNSOpenSSL/shim.h]: Thanks, I know its pedantic, but this is the world we live in with dual compilation platforms and 1000 versions of OpenSSL on each.

Once SSWG gets it's own version of OpenSSL lib we can remove this and be done with it.

File: Tests/NIOAPNSJWTTests/JWTTests.swift (hunk 6 above)
Comment (kylebrowning) [reply Tests/NIOAPNSJWTTests/JWTTests.swift]: Are you planning on doing this now or should I merge and do it later?

Comment (lgaches) [reply Tests/NIOAPNSJWTTests/JWTTests.swift]: I'm planning to add more tests. But maybe you can merge and I do it later in a new PR.

Comment (kylebrowning) [reply Tests/NIOAPNSJWTTests/JWTTests.swift]: Sounds good. thank you so much for this work, this is a HUGE addition!
